src/main/java/org/opensearch/ad/ratelimit/RateLimitedQueue.java

```diff
@@ -0,0 +1,474 @@
+                    setCoolDownStart();
+                    break;
+                }
+            }
+        }
+    }
+
+    private void prune(Map<String, Segment> segments) {
+        for (Map.Entry<String, Segment> segmentEntry : segments.entrySet()) {
+            if (segmentEntry.getKey().equals(SegmentPriority.HIGH.name())) {
+                return;
+            }
+            // remove more requests in the low priority segment
+            float removeRatio = mediumSegmentPruneRatio;
+            if (segmentEntry.getKey().equals(SegmentPriority.LOW.name())) {
+                removeRatio = lowSegmentPruneRatio;
+            }
+
+            Segment segment = segmentEntry.getValue();
+
+            if (segment == null) {
+                return;
+            }
+
+            BlockingQueue<RequestType> segmentContent = segment.content;
+            // remove 10% of old requests
+            int deletedRequests = (int) (segmentContent.size() * removeRatio);
+            while (false == segmentContent.isEmpty() && deletedRequests-- >= 0) {
+                segmentContent.poll();
+            }
+        }
+    }
+
+    private void maintainForMemory() {
+        // removed expired segment
+        maintenance(requestSegments, stateTtl);
+
+        if (isSizeExceeded()) {
+            // remove until reaching below queueSize
+            do {
+                prune(requestSegments);
+            } while (isSizeExceeded());
+        } else if (adCircuitBreakerService.isOpen()) {
+            // remove a few items in each segment
+            prune(requestSegments);
+        }
+    }
+
+    private boolean isSizeExceeded() {
+        Collection<Segment> segments = requestSegments.values();
+        int totalSize = 0;
+
+        // When faced with a backlog beyond the limit, we prefer fresh requests
+        // and throws away old requests.
+        // release space so that put won't block
+        for (Segment segment : segments) {
+            totalSize += segment.size();
+        }
+        return totalSize >= queueSize;
+    }
+
+    @Override
+    public void maintenance() {
+        try {
+            maintainForMemory();
+            maintainForThreadPool();
+        } catch (Exception e) {
+            // it is normal to maintain why other
+            LOG.warn("Failed to maintain", e);
+        }
+    }
+
+    /**
+     * Start cooldown during a overloaded situation
+     */
+    protected void setCoolDownStart() {
+        cooldownStart = clock.instant();
+    }
+
+    /**
+     * @param batchSize the max number of requests to fetch
+     * @return a list of batchSize requests (can be less)
+     */
+    protected List<RequestType> getRequests(int batchSize) {
+        List<RequestType> toProcess = new ArrayList<>(batchSize);
+
+        Set<BlockingQueue<RequestType>> selectedQueue = new HashSet<>();
+
+        while (toProcess.size() < batchSize) {
```

Comment:
Won't toProcess initially have a size of batchSize?

Reply:
it won't:

jshell> List<Integer> a = new ArrayList<>(5);
a ==> []

jshell> a.size()
$2 ==> 0